owncloudApp/src/main/java/com/owncloud/android/authentication/AccountAuthenticator.java

```diff
@@ -306,69 +305,87 @@ private boolean canBeRefreshed(String authTokenType) {
 
-            OwnCloudClient client = OwnCloudClientFactory.createOwnCloudClient(
-                    Uri.parse(accountManager.getUserData(account, AccountUtils.Constants.KEY_OC_BASE_URL)),
-                    mContext,
-                    true
-            );
+        // Prepare everything to perform the token request
 
-            RemoteOperationResult<Map<String, String>> result = oAuth2RefreshAccessTokenOperation.execute(client);
-            if (!result.isSuccess()) {
-                Timber.e("Failed to refresh access token");
-                return null;
-            }
-
-            // Get new access and refresh tokens
-            Map<String, String> tokens = result.getData();
-
-            accessToken = tokens.get(OAuth2Constants.KEY_ACCESS_TOKEN);
-
-            Timber.d("Got OAuth2 access token: %s", accessToken);
-
-            accountManager.setAuthToken(account, authTokenType, accessToken);
+        String refreshToken = accountManager.getUserData(
+                account,
+                AccountUtils.Constants.KEY_OAUTH2_REFRESH_TOKEN
+        );
 
-            Timber.d("Set OAuth2 new access token in account: %s", accessToken);
-
-            refreshToken = tokens.get(OAuth2Constants.KEY_REFRESH_TOKEN);
-
-            Timber.d("Got OAuth2 refresh token: %s", refreshToken);
-
-            accountManager.setUserData(
-                    account,
-                    AccountUtils.Constants.KEY_OAUTH2_REFRESH_TOKEN,
-                    refreshToken
-            );
-
-        } catch (Exception e) {
-            Timber.e(e, "Failed to store refreshed access token");
-            return null;
+        if (refreshToken == null || refreshToken.length() <= 0) {
+            Timber.w("No refresh token stored for silent renewal of access token");
+            return;
         }
 
-        return accessToken;
+        Timber.d("Get OAuth2 refresh token from account: %s, to exchange it for new access and refresh tokens",
+                refreshToken);
+
+        String baseUrl = accountManager.getUserData(
+                account,
+                AccountUtils.Constants.KEY_OC_BASE_URL
+        );
+
+        AuthorizationServiceConfiguration serviceConfiguration = new AuthorizationServiceConfiguration(
+                Uri.parse(baseUrl + "/" + mContext.getString(R.string.oauth2_url_endpoint_auth)), // auth endpoint
+                Uri.parse(baseUrl + "/" + mContext.getString(R.string.oauth2_url_endpoint_access)) // token endpoint
+        );
+
+        String scope = accountManager.getUserData(
+                account,
+                AccountUtils.Constants.KEY_OAUTH2_SCOPE
+        );
+
+        TokenRequest tokenRequest = new TokenRequest.Builder(
+                serviceConfiguration,
+                mContext.getString(R.string.oauth2_client_id))
+                .setGrantType(GrantTypeValues.REFRESH_TOKEN)
+                .setScope(scope)
+                .setRefreshToken(refreshToken)
+                .build();
```

Comment:
Yes, though it could also be indented as:

```
TokenRequest tokenRequest = new TokenRequest.Builder(
    serviceConfiguration,
    mContext.getString(R.string.oauth2_client_id)
).setGrantType(GrantTypeValues.REFRESH_TOKEN)
    .setScope(scope)
    .setRefreshToken(refreshToken)
    .build();
```

Reply:
I think it is easier to read that way, thanks